workspace: Add function to save new file in directory nearest tab

CharlesChen0823 opened the pull request:
Closes #15685

Release Notes:

- Save dialogues for new files will now default to the directory you're currently working in instead of the project root.

CharlesChen0823 committed:
init

CharlesChen0823 commented:
> There is still an edge case, when save file in split pane with no nearest tab behaviour not save with vscode.

Don't have find better method for this edge case, Does someone have better idea?

CharlesChen0823 committed:
add

ConradIrwin commented:
@CharlesChen0823 Thanks for this!

I think it would be better to make this the default (like Sublime) than to have two different approaches for the same thing.

CharlesChen0823 commented:
> I think it would be better to make this the default (like Sublime) than to have two different approaches for the same thing.

I'm not understand what's your meaning?

CharlesChen0823 committed:
Merge remote-tracking branch 'upstream/main' into save_with_nearest

CharlesChen0823 committed:
fix

ConradIrwin commented:
@CharlesChen0823 sorry for dropping this. I had misunderstood your change. I think this is an improvement, and we should merge it. Do you want to fix the merge conflicts?

CharlesChen0823 committed:
Merge remote-tracking branch 'upstream/main' into save_with_nearest

CharlesChen0823 committed:
fix merge

CharlesChen0823 committed:
format

CharlesChen0823 commented:
the CI failed is fakely, and current status can be review. 
But there is one edge case:
1. open two pane, left and right.
2. open one file in left one.
3. create new file in right one
4. save new file, will not get focus left one directory.

Because `last_active_center_pane` most time same with `active_pane`.

I want fixed `last_active_center_pane` with 
```diff
diff --git a/crates/workspace/src/workspace.rs b/crates/workspace/src/workspace.rs
index 813a03488d..bcb5438f4e 100644
--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -2859,12 +2859,13 @@ impl Workspace {
         cx: &mut App,
     ) -> Task<Result<Box<dyn ItemHandle>, anyhow::Error>> {
         let pane = pane.unwrap_or_else(|| {
-            self.last_active_center_pane.clone().unwrap_or_else(|| {
-                self.panes
-                    .first()
-                    .expect("There must be an active pane")
-                    .downgrade()
-            })
+            self.active_pane.downgrade()
+            // self.last_active_center_pane.clone().unwrap_or_else(|| {
+            //     self.panes
+            //         .first()
+            //         .expect("There must be an active pane")
+            //         .downgrade()
+            // })
         });
 
         let task = self.load_path(path.into(), window, cx);
@@ -2906,10 +2907,11 @@ impl Workspace {
         cx: &mut Context<Self>,
     ) -> Task<Result<Box<dyn ItemHandle>, anyhow::Error>> {
         let pane = self.last_active_center_pane.clone().unwrap_or_else(|| {
-            self.panes
-                .first()
-                .expect("There must be an active pane")
-                .downgrade()
+            self.active_pane.downgrade()
+            // self.panes
+            //     .first()
+            //     .expect("There must be an active pane")
+            //     .downgrade()
         });
 
         if let Member::Pane(center_pane) = &self.center.root {
@@ -3357,9 +3359,10 @@ impl Workspace {
         window: &mut Window,
         cx: &mut Context<Self>,
     ) {
+        self.last_active_center_pane = Some(self.active_pane.downgrade());
         self.active_pane = pane.clone();
         self.active_item_path_changed(window, cx);
-        self.last_active_center_pane = Some(pane.downgrade());
+        // self.last_active_center_pane = Some(pane.downgrade());
     }
 
     fn handle_panel_focused(&mut self, window: &mut Window, cx: &mut Context<Self>) {
```
But I'm not sure if there any other side effect.

CharlesChen0823 committed:
using new variable to record last_active_pane

ConradIrwin committed:
Merge branch 'main' into save_with_nearest

ConradIrwin committed:
Use existing history for implementation.

Also avoid the dance of reading it on workspace, passing it down to
pane, and then passing it back to workspace, by reading it in the async
hop.

ConradIrwin commented:
@CharlesChen0823 I pushed up a change to the implementation that makes it closer match Sublime Text. Instead of trying to figure things out based on adjacent items in the panes, we just find the most recent history item, and use that.

Also, given that the workspace has all the information it needs to do this in `.prompt_for_new_path` already, I updated things so that method does this directly, avoiding the need to change `Pane::save_item`.

Thanks again for getting the ball rolling on this, and sorry it took so long to get to a merge point!
